### ISSUE TITLE
Spaces after « and before » should be nobreak in french

### ISSUE THREAD

**Issue by crowdagger** [typst/typst#1920]
### Description

(Hi, first time submitting an issue for this project, hope I'm doing it things more or less right!)

When setting lang to french, spaces after and before the french « quotes » (or "guillemets") « and » should be unbreakable. Currently, this is not the case: 

```
« Bonjour, ceci est un dialogue » 

"Bonjour, ceci est un dialogue"
```
![Screenshot 2023-08-12 at 14-08-44 Test – Typst](https://github.com/typst/typst/assets/1961791/6957e2c6-0e7f-404b-8323-32d5648da161)

As the example above shows, using the smart quote feature it works correctly, but some french layouts (e.g. Bépo) allow to easily type « and » and if you are used to do it the results won't be great.

(If you agree this should be fixed and can point me in the right direction, I'd be happy to try and start working on a PR ^^)

### Operating system

Web app, Windows

### Typst version

- [X] I am using the latest version of Typst

**Comment by Omikhleia** [comment]
Slightly relates to https://github.com/typst/typst/issues/1194#issuecomment-1561194499 (where I mentioned in passing that "the use of a `\u{00A0}` space in French quotes is not really respecting the true typography")

My 2 cents on the topic:

The "guillemet space" shall always be unbreakable indeed, and is "usually" smaller than the regular inter-word space, with also a bit less stretch/shrinkability. Babel (LaTeX) adopted a definition recommended by Thierry Bouche years ago (and SILE followed it too more recently), which as far as I understand is typographically sound for `fr` and `fr-CA` (but note that `fr-CH` may prefer a thin space there, or even no guillemet space at all).

Likewise, for the record, French requires a non-breakable space (but usually stretch/shrinkable as an inter-word space for justification purposes) before a colon -- and a thin space before certain "high" punctuation signs `;!?` (with a number of exceptions...), unbreakable too and usually non-variable, ideally around half of the inter-word space, at least in `fr`; `fr-CA` usually does not add that space before the so-said high punctuations)

Obviously/unfortunately, few typesetting software follow these usual/old conventions -- as noted, LaTeX with Babel does, as well as SILE -- But many just use some unbreakable _fixed_ spaces in these cases. FWIW, LibreOffice and MSWord do that (i.e. insert the same fixed nbsp everywhere), though they do implement some of the differences between French from France, Canada and Switzerland ;)

**Comment by yodatak** [comment]
Maybe in this part https://github.com/typst/typst/blob/main/crates/typst/src/text/smartquote.rs

any maybe with this we could now doi it with simple show instruction
https://github.com/typst/typst/pull/4876
https://github.com/typst/typst/issues/4672

**Comment by laurmaedje** [comment]
Not sure whether this is even a bug. The smartquotes work as intended and the text above with manual quotes simply doesn't use non-breaking spaces.

**Comment by knuesel** [comment]
I agree this is rather a feature request (the current behavior is fine but maybe typst could automatically fix the spacing when the use enters guillemets).

If this issue is closed though maybe #4674 should be reopened (it would be nice to extend the "automatic localization" done for quotes to other types of punctuation).

**Comment by ainar** [comment]
I'm facing this problem in the bibliography. I know no way to add a space before a colon in this case:
![image](https://github.com/user-attachments/assets/e2d75562-7627-422f-8a71-870127d561e5)

**Comment by eltos** [comment]
How about a simple show rule to ease your typing habit?

```typ
#set page(width: 190pt, height: auto)

« Bonjour, ceci est un dialogue » 

#show regex("(« |‹ | ›| »)"): it => {
  show sym.space: sym.space.nobreak
  it
}

« Bonjour, ceci est un dialogue » 
```

![testsdf.png](https://github.com/user-attachments/assets/f9b92e73-83c0-4500-b185-bd512ffb7068)



Note that the usage of Guillemets differs in different cultures, so automatic space substitution would have to take these into account:
```
« Beispiel in Frankreich und der Welschschweiz »
«Beispiel in der Deutschschweiz»
»Beispiel in Deutschland und Österreich«
```

**Comment by laurmaedje** [comment]
Fixed by https://github.com/typst/typst/pull/5976

**Comment by MDLC01** [comment]
#5976 is about the smartquote elements. This issue is about manually inserted guillemets. The following still does not work as intended in the April 10 development version, which should include #5976.

```typ
#set text(lang: "fr")

#lorem(14) A test » \
#lorem(14) A test ! \
#lorem(14) A test ? \
#lorem(14) A test : \
#lorem(14) A test ; \
#lorem(14) A test ,
```

![Only the closing guillemet is alone on a new line. The other marks are on a new line, together with "test".](https://github.com/user-attachments/assets/390cda60-eef1-44d8-a102-4b32f5e6db49)

**Comment by Enivex** [comment]
If you're manually inserting quotes, should you then not also manually insert no-break spaces?

**Comment by MDLC01** [comment]
As I explained in https://github.com/typst/typst/issues/1920#issuecomment-2679535824, spaces before some punctuation marks already do behave as non-breaking spaces.

**Comment by laurmaedje** [comment]
> https://github.com/typst/typst/pull/5976 is about the smartquote elements. This issue is about manually inserted guillemets. The following still does not work as intended in the April 10 development version, which should include https://github.com/typst/typst/pull/5976.

I was a bit too fast with this one ...

**Comment by MrVyM** [comment]
I'm currently working on a document on typst with smartquotes. I just find this. 
It's the correct place to talk about this ? 
<img width="258" height="110" alt="Image" src="https://github.com/user-attachments/assets/43dae5ca-f614-48d9-82c9-717e43d1cf5d" />

<img width="313" height="111" alt="Image" src="https://github.com/user-attachments/assets/ba4dbb6d-b62c-4fe9-861f-70505936afa7" />

**Comment by MDLC01** [comment]
You may want to create a separate issue, but in this case I don't think this behavior is unintended.

**Comment by laurmaedje** [comment]
Closed by https://github.com/typst/typst/pull/8406.
